internal/pkg/earpointer/analysis.go

```diff
@@ -0,0 +1,649 @@
+		vis.visitConvert(i)
+	case *ssa.ChangeInterface:
+		vis.visitChangeInterface(i)
+	case *ssa.ChangeType:
+		vis.visitChangeType(i)
+	case *ssa.TypeAssert:
+		vis.visitTypeAssert(i)
+	case *ssa.UnOp:
+		vis.visitUnOp(i)
+	case *ssa.BinOp:
+		vis.visitBinOp(i)
+	case *ssa.Extract:
+		vis.visitExtract(i)
+	case *ssa.Range:
+		vis.visitRange(i)
+	case *ssa.Send:
+		vis.visitSend(i)
+	case *ssa.Slice:
+		vis.visitSlice(i)
+	case *ssa.MakeInterface:
+		vis.visitMakeInterface(i)
+	case *ssa.Select:
+		vis.VisitSelect(i)
+
+	case *ssa.Call:
+		vis.visitCall(i.Call, instr)
+	case *ssa.Go:
+		vis.visitCall(i.Call, instr)
+	case *ssa.Defer:
+		vis.visitCall(i.Call, instr)
+
+	case *ssa.Alloc,
+		*ssa.MakeChan,
+		*ssa.MakeClosure,
+		*ssa.MakeMap,
+		*ssa.MakeSlice:
+		// Local variables are created in the beginning of the pass by initReferences().
+		// These instructions won't be handled after that.
+	case *ssa.If,
+		*ssa.Jump:
+		// The analysis is flow insensitive.
+	case *ssa.Next:
+		// Next is handled in visitExtract().
+	case *ssa.Return:
+		// Return is handled in visitCall().
+	case *ssa.RunDefers:
+		// To be implemented.
+	case *ssa.Panic:
+		// To be implemented.
+	case *ssa.DebugRef:
+		// Not related.
+	}
+}
+
+func (vis *visitor) visitFieldAddr(addr *ssa.FieldAddr) {
+	if !typeMayShareObject(addr.Type()) {
+		return
+	}
+	// "t1 = t0.x" is implemented by t0[x -> t1] through address unification.
+	obj := addr.X
+	field := makeField(obj.Type(), addr.Field)
+	for _, c := range vis.contexts {
+		vis.unifyFieldAddress(c, obj, field, addr)
+	}
+}
+
+func (vis *visitor) visitField(field *ssa.Field) {
+	// "t1 = t0.x" is implemented by t0[x -> t1].
+	obj := field.X
+	fd := makeField(obj.Type(), field.Field)
+	vis.processHeapAccess(field, obj, *fd)
+}
+
+func (vis *visitor) visitIndexAddr(addr *ssa.IndexAddr) {
+	// "t2 = &t1[t0]" is handled through address unification.
+	vis.processMemoryIndexAccess(addr, addr.X, addr.Index)
+}
+
+func (vis *visitor) visitIndex(index *ssa.Index) {
+	vis.processMemoryIndexAccess(index, index.X, index.Index)
+}
+
+func (vis *visitor) visitLookup(lookup *ssa.Lookup) {
+	if !lookup.CommaOk {
+		vis.processMemoryIndexAccess(lookup, lookup.X, lookup.Index)
+	}
+	// The comma case will be processed by the subsequent extract instruction.
+}
+
+func (vis *visitor) visitStore(store *ssa.Store) {
+	// Here store instruction "*address = data" is implemented by a
+	// semantics-equivalent (w.r.t. unification) load instruction "data =
+	// *address".
+	// The address can be a (memory) address, a global variable, or a free
+	// variable.
+	if mayShareObject(store.Val) {
+		vis.processAddressToValue(store.Addr, store.Val)
+	}
+}
+
+func (vis *visitor) visitPhi(phi *ssa.Phi) {
+	if !typeMayShareObject(phi.Type()) {
+		return
+	}
+	for _, e := range phi.Edges {
+		vis.unifyLocals(phi, e)
+	}
+}
+
+func (vis *visitor) visitMapUpdate(update *ssa.MapUpdate) {
+	vis.processMemoryIndexAccess(update.Value, update.Map, update.Key)
+}
```

Comment:
Nit: consider moving to group with other `processMemoryIndexAccess` delegators.

Reply:
The handling of these instructions may be different when we add more login into the visitor, e.g. when object sensitivity is supported. I will keep them separate for the moment.